### ISSUE TITLE
How to release gpu memory without exiting the process？

### ISSUE THREAD

**Issue by feipxyz** [microsoft/onnxruntime#7463]
after run these code，how to free gpu memory manually without exiting the process？
```
 ort_inputs = {self.ort_session.get_inputs()[0].name: images}
 ort_outs = self.ort_session.run(None, ort_inputs)
```

**System information**
- Ubuntu 20.04:
- pip install onnxruntime-gpu:
- onnxruntime-gpu 1.7.0:
- Python3.8
- CUDA/cuDNN version:
- cuda11.0 cudnn11.2

**Comment by skottmckay** [comment]
We use a memory pool for the GPU memory. That is freed when the ORT session is deleted. 

Currently there's no mechanism to explicitly free memory that the session is using whilst keeping the session around.

**Comment by feipxyz** [comment]
> We use a memory pool for the GPU memory. That is freed when the ORT session is deleted.
> 
> Currently there's no mechanism to explicitly free memory that the session is using whilst keeping the session around.

I want to limit memory usage when forward. I refer to this document[CUDA Execution Provider](https://www.onnxruntime.ai/docs/reference/execution-providers/CUDA-ExecutionProvider.html). 
```python
so = onnxruntime.SessionOptions()
        so.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_ENABLE_ALL # ORT_ENABLE_EXTENDED ORT_ENABLE_ALL
        so.intra_op_num_threads = 4
        so.execution_mode = onnxruntime.ExecutionMode.ORT_SEQUENTIAL # ORT_PARALLEL ORT_SEQUENTIAL
        providers = [
            ('CUDAExecutionProvider', {
                'device_id': 0,
                'arena_extend_strategy': 'kNextPowerOfTwo',
                'cuda_mem_limit': 4 * 1024 * 1024 * 1024,
                'cudnn_conv_algo_search': 'EXHAUSTIVE',
                'do_copy_in_default_stream': True,
            }),
            'CPUExecutionProvider',
        ]
        # self.ort_session = onnxruntime.InferenceSession(model_path, sess_options=so)              
        self.ort_session = onnxruntime.InferenceSession(model_path, sess_options=so, providers=providers)             
```
But An error occurred during running 
```
    self.ort_session = onnxruntime.InferenceSession(model_path, sess_options=so, providers=providers)
  File "C:\Users\1\miniconda3\envs\py37\lib\site-packages\onnxruntime\capi\session.py", line 195, in __init__
    self._create_inference_session(providers, provider_options)
  File "C:\Users\1\miniconda3\envs\py37\lib\site-packages\onnxruntime\capi\session.py", line 205, in _create_inference_session
    sess.initialize_session(providers or [], provider_options or [])
TypeError: initialize_session(): incompatible function arguments. The following argument types are supported:
    1. (self: onnxruntime.capi.onnxruntime_pybind11_state.InferenceSession, arg0: List[str], arg1: List[Dict[str, str]]) -> None

Invoked with: <onnxruntime.capi.onnxruntime_pybind11_state.InferenceSession object at 0x000001DD895D80F0>, [('CUDAExecutionProvider', {'device_id': 0, 'arena_extend_strategy': 'kNextPowerOfTwo', 'cuda_mem_limit': 4294967296, 'cudnn_conv_algo_search': 'EXHAUSTIVE', 'do_copy_in_default_stream': True}), 'CPUExecutionProvider'], []
```
How to do it right？

**Comment by skottmckay** [comment]
I pasted your `providers` config and tried with ORT v1.7 with no problem. Can you check the ORT version you're using in the miniconda environment (`help(onnxruntime)` after importing onnxruntime should print the version near the end of the output) to make sure it's 1.7?

**Comment by arch1902** [comment]
> We use a memory pool for the GPU memory. That is freed when the ORT session is deleted.
> 
> Currently there's no mechanism to explicitly free memory that the session is using whilst keeping the session around.

I tried deleting the ort_session using "del ort_session" but the GPU still shows memory usage. I want to clear GPU memory for other models within the same python process ?

**Comment by vahvero** [comment]
@arch1902 Have you tried manually calling garbace collection with `import gc;gc.collect()` after `del ort_session`? I have similar issues in my tests where the memory is not freed and they fail with memory allocation error. However, the tests pass if ran singularly with process exit in between.

**Comment by arch1902** [comment]
@vahvero Yes, I did try that but it doesn't reduce the memory in use on GPU.

**Comment by vahvero** [comment]
Do you have any calls with `functools.cache` or similar structure where session might be cached?

**Comment by arch1902** [comment]
@vahvero  I am using a simple python function so I hope there is no caching involved. My guess is CUDA itself is caching some memory. Also, I am using just `nvidia-smi` to monitor memory usage which might not be the most detailed one.

**Comment by hyperparameters** [comment]
I am having a similar issue, did got any solution ?@arch1902

**Comment by vahvero** [comment]
If it is a simple function, I could try to replicate it if you are willing to share.

**Comment by ibrahimsoliman97** [comment]
Any update on this issue?

**Comment by m-evdokimov** [comment]
I solved the issue by initializing ort session on every inference call. Works well for my purposes

**Comment by panxin801** [comment]
@m-evdokimov I face the same problem, and I want to try your solution, but Create new ort Session each time may cause extra time consumation I think, have you ever test the efficiency?

**Comment by m-evdokimov** [comment]
@panxin801 Hi! Turns out it takes unacceptable amount of time to optimize model (even if you use graph optimizations beforehand, it'll take extra time to make other optimizations that's hard to adjust before initialization). My guess is that for now this is the only way to release the memory.

**Comment by panxin801** [comment]
> @panxin801 Hi! Turns out it takes unacceptable amount of time to optimize model (even if you use graph optimizations beforehand, it'll take extra time to make other optimizations that's hard to adjust before initialization). My guess is that for now this is the only way to release the memory.

Well, thansks for reply. Yeah, this may be the only way to release the memory . Thanks mate.
